suite/tests/linux/clone.c

```diff
@@ -106,24 +109,39 @@ test_thread(bool share_sighand, bool clone_vm, bool use_clone3)
 {
     print("%s(share_sighand %d, clone_vm %d, use_clone3 %d)\n", __FUNCTION__,
           share_sighand, clone_vm, use_clone3);
-    if (use_clone3) {
-#ifdef SYS_clone3
+
+    if (use_clone3 && clone3_available)
         child = create_thread_clone3(run_with_exit, &stack, share_sighand, clone_vm);
-#else
-        /* If SYS_clone3 is not defined, we simply use SYS_clone instead, so that
-         * the expected output is the same in both cases.
+    else {
+        /* If use_clone3 is true but clone3 is not available, we simply use
+         * SYS_clone instead, so that the expected output is the same.
          */
         child = create_thread(run, NULL, &stack, share_sighand, clone_vm);
-#endif
-    } else
-        child = create_thread(run, NULL, &stack, share_sighand, clone_vm);
+    }
+
     assert(child > -1);
     delete_thread(child, stack);
 }
 
 int
 main()
 {
+    /* Try using clone3 when it is possibly not defined. */
+    int ret_failure_clone3 = make_clone3_syscall(NULL, 0, NULL);
+    assert(ret_failure_clone3 == -1);
+#ifdef SYS_clone3
+    /* In some scenarios SYS_clone3 is defined but clone3 returns ENOSYS
+    * e.g. when running in a container under Ubuntu 22.04
+    * see https://github.com/moby/moby/pull/42681
+    */
+    if (errno != ENOSYS)
+        clone3_available = true;
```

Comment:
IIUC, we're only setting clone_available to true in the case where SYS_clone3 is defined.
The code can handle clone3 being supported even if SYS_clone3 is not defined, do we want to handle that here?

Reply:
Have removed the `#ifdef SYS_clone3` so that `clone3` can be tested when SYS_clone3 is not defined.